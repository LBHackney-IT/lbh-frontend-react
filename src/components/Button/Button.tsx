import classNames from "classnames";
import GovukButton from "govuk-frontend/govuk/components/button/button";
import { nullValuesAsUndefined } from "null-as-undefined";
import PropTypes from "prop-types";
import React from "react";

import { StartButtonIcon } from "../icons/StartButtonIcon";

import "lbh-frontend/lbh/components/lbh-button/_button.scss";

/**
 * The proptypes for {@link Button}.
 */
export interface ButtonProps {
  id?: string | null;
  className?: string | null;
  /**
   * Use for the main call to action on your service's start page. Adjusts
   * style of the button and adds a start icon as the last child.
   */
  isStartButton?: boolean | null;
  /**
   * Debounces clicks within a second of each other preventing multiple form
   * submissions.
   */
  preventDoubleClick?: boolean | null;
  name?: string | null;
  type?: React.ButtonHTMLAttributes<HTMLButtonElement>["type"] | null;
  /**
   * The text of the button.
   *
   * This is followed by the start icon if {@link ButtonProps.isStartButton} is
   * `true`.
   */
  children: React.ReactNode;
  disabled?: boolean | null;
  onClick?: ((event: React.MouseEvent<HTMLElement>) => void) | null;
}

/**
 * The button element, that can be used to help users
 * carry out an action.
 */
export class Button extends React.Component<ButtonProps> {
  static propTypes: PropTypes.ValidationMap<ButtonProps> = {
    id: PropTypes.string,
    className: PropTypes.string,
    isStartButton: PropTypes.bool,
    preventDoubleClick: PropTypes.bool,
    name: PropTypes.string,
    type: PropTypes.oneOf(["submit", "button", "reset"]),
    children: PropTypes.node.isRequired,
    disabled: PropTypes.bool,
    onClick: PropTypes.func
  };

  // This is used by `Button.prototype.debounce`.
  private debounceFormSubmitTimer: NodeJS.Timeout | null = null;

  private handleClick(event: React.MouseEvent<HTMLElement>): void {
    const { onClick, disabled } = this.props;

    if (disabled) {
      return;
    }

    // `debounce` calls `event.preventDefault()` for us, so we don't need to
    // repeat that behaviour.
    const shouldHandleClick = GovukButton.prototype.debounce.call(this, event);

    if (shouldHandleClick === false) {
      return;
    }

    if (onClick) {
      onClick(event);
    }
  }

  render(): JSX.Element {
    const {
      id,
      className,
      isStartButton,
      preventDoubleClick,
      name,
      type,
      children,
      disabled
    } = nullValuesAsUndefined(this.props);
    return (
      <button
        id={id}
        className={classNames("govuk-button lbh-button", className, {
          "govuk-button--start": isStartButton
        })}
        name={name}
        type={type}
        disabled={disabled}
        aria-disabled={disabled}
        data-prevent-double-click={preventDoubleClick}
        onClick={this.handleClick.bind(this)}
      >
        {children}
        {isStartButton && (
          <StartButtonIcon className="govuk-button__start-icon" />
        )}
      </button>
    );
  }
}
