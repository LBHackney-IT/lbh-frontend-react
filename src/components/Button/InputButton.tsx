import classNames from "classnames";
import { nullValuesAsUndefined } from "null-as-undefined";
import PropTypes from "prop-types";
import React from "react";

import { debounce, makeDebounceContext } from "../../helpers/debounce";
import "lbh-frontend/lbh/components/lbh-button/_button.scss";

/**
 * The proptypes for {@link InputButton}.
 */
export interface InputButtonProps {
  id?: string | null;
  className?: string | null;
  /**
   * Use for the main call to action on your service's start page. Adjusts the
   * styling of the button but, unlike {@link Button}, does not add an icon due
   * to this using an `<input>` internally.
   */
  isStartButton?: boolean | null;
  /**
   * Debounces clicks within a second of each other preventing multiple form
   * submissions.
   */
  preventDoubleClick?: boolean | null;
  name?: string | null;
  type: React.ButtonHTMLAttributes<HTMLButtonElement>["type"];
  value?: string | null;
  disabled?: boolean | null;
}

/**
 * A component for creating an input element, styled as a button. It can be
 * used to help users carry out actions such as submitting forms.
 */
export class InputButton extends React.Component<InputButtonProps> {
  static propTypes: PropTypes.ValidationMap<InputButtonProps> = {
    id: PropTypes.string,
    className: PropTypes.string,
    isStartButton: PropTypes.bool,
    preventDoubleClick: PropTypes.bool,
    name: PropTypes.string,
    type: PropTypes.oneOf<"submit" | "button" | "reset">([
      "submit",
      "button",
      "reset"
    ]).isRequired,
    value: PropTypes.string,
    disabled: PropTypes.bool
  };

  private debounceContext = makeDebounceContext();

  private handleClick(event: React.MouseEvent<HTMLInputElement>): void {
    const { disabled } = this.props;

    if (disabled) {
      event.preventDefault();
      return;
    }

    if (debounce(event, this.debounceContext)) {
      return;
    }
  }

  /**
   * @ignore
   */
  render(): JSX.Element {
    const {
      id,
      className,
      isStartButton,
      preventDoubleClick,
      name,
      type,
      value,
      disabled
    } = nullValuesAsUndefined(this.props);
    return (
      <input
        id={id}
        className={classNames("govuk-button lbh-button", className, {
          "govuk-button--start": isStartButton
        })}
        name={name}
        value={value}
        type={type}
        disabled={disabled}
        aria-disabled={disabled}
        onClick={this.handleClick.bind(this)}
        data-prevent-double-click={preventDoubleClick}
      />
    );
  }
}
