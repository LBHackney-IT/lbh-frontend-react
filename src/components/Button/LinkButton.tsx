import classNames from "classnames";
import GovukButton from "govuk-frontend/govuk/components/button/button";
import PropTypes from "prop-types";
import React, { Component } from "react";

import { nullAsUndefined } from "../../helpers/nullAsUndefined.internal";
import { ComponentRegister } from "../../helpers/ComponentRegister";
import { StartButtonIcon } from "../icons/StartButtonIcon";

import "lbh-frontend/lbh/components/lbh-button/_button.scss";

/**
 * The proptypes for {@link LinkButtonComponentType}.
 */
export interface LinkButtonProps {
  id?: string | null;
  className?: string | null;
  /**
   * Use for the main call to action on your service's start page. Adjusts
   * style of button and adds a start icon.
   */
  isStartButton?: boolean | null;
  href: string;
  children: React.ReactNode;
}

/**
 * A component for creating an anchor linking to other pages, styled as a
 * button. You can customize the component used internally for integration with
 * routers using {@link ComponentRegister.init}. By default, it uses an `<a>`
 * tag.
 */
export class LinkButton extends Component<LinkButtonProps> {
  static propTypes: PropTypes.ValidationMap<LinkButtonProps> = {
    id: PropTypes.string,
    className: PropTypes.string,
    isStartButton: PropTypes.bool,
    href: PropTypes.string.isRequired,
    children: PropTypes.node.isRequired
  };

  /**
   * @ignore
   */
  render(): JSX.Element {
    const { id, className, isStartButton, href, children } = this.props;
    return (
      <ComponentRegister.LinkButton
        id={nullAsUndefined(id)}
        className={classNames("govuk-button lbh-button", className, {
          "govuk-button--start": isStartButton
        })}
        href={href}
        role="button"
        onKeyDown={GovukButton.prototype.handleKeyDown.bind(this)}
      >
        {children}
        {isStartButton && <StartButtonIcon />}
      </ComponentRegister.LinkButton>
    );
  }
}
