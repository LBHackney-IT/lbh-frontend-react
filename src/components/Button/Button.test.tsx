import { mount } from "enzyme";
import React from "react";
import { create } from "react-test-renderer";
import { Button } from "./Button";

it("renders correctly with all props", () => {
  const component = create(
    <Button
      id="1234"
      className="class1 class2"
      isStartButton
      preventDoubleClick
      name="Testname"
      type="button"
      disabled
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      onClick={(): void => {}}
      aria-label="test button"
      data-test="test data"
    >
      Testchildren
    </Button>
  );

  expect(component).toMatchInlineSnapshot(`
    <button
      aria-disabled={true}
      aria-label="test button"
      className="govuk-button lbh-button class1 class2 govuk-button--start"
      data-prevent-double-click={true}
      data-test="test data"
      disabled={true}
      id="1234"
      name="Testname"
      onClick={[Function]}
      type="button"
    >
      Testchildren
      <svg
        className="govuk-button__start-icon"
        focusable="false"
        height="19"
        role="presentation"
        viewBox="0 0 33 40"
        width="17.5"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h13l20 20-20 20H0l20-20z"
          fill="currentColor"
        />
      </svg>
    </button>
  `);
});

it("renders correctly without optional props", () => {
  const component = create(<Button>Testchildren</Button>);

  expect(component).toMatchInlineSnapshot(`
    <button
      className="govuk-button lbh-button"
      onClick={[Function]}
    >
      Testchildren
    </button>
  `);
});

it("calls the onClick prop when clicked", () => {
  const mockClickHandler = jest.fn();

  const wrapper = mount(
    <Button
      id="1234"
      className="class1 class2"
      preventDoubleClick
      name="Testname"
      type="button"
      onClick={mockClickHandler}
    >
      Testchildren
    </Button>
  );

  const button = wrapper.find("button");
  const buttonOnClick = button.props().onClick;

  expect(buttonOnClick).toBeDefined();

  const mockPreventDefault = jest.fn();
  const mockClickEvent = {
    ...jest.fn()(),
    target: button.getDOMNode(),
    preventDefault: mockPreventDefault,
  };

  //eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  buttonOnClick!.call(button, mockClickEvent);

  expect(mockClickHandler).toHaveBeenCalledTimes(1);
  expect(mockClickHandler).toHaveBeenCalledWith(mockClickEvent);
  expect(mockPreventDefault).not.toHaveBeenCalled();
});

it("prevents a second click within a second from executing the button a second time when preventDoubleClick is true", () => {
  const mockClickHandler = jest.fn();

  const wrapper = mount(
    <Button preventDoubleClick onClick={mockClickHandler}>
      Testchildren
    </Button>
  );

  const button = wrapper.find("button");
  const buttonOnClick = button.props().onClick;

  expect(buttonOnClick).toBeDefined();

  const mockPreventDefault = jest.fn();
  const mockClickEvent = {
    ...jest.fn()(),
    target: button.getDOMNode(),
    preventDefault: mockPreventDefault,
  };

  //eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  buttonOnClick!.call(button, mockClickEvent);

  expect(mockClickHandler).toHaveBeenCalledTimes(1);
  expect(mockClickHandler).toHaveBeenCalledWith(mockClickEvent);
  expect(mockPreventDefault).not.toHaveBeenCalled();

  mockClickHandler.mockReset();

  jest.advanceTimersByTime(999);

  //eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  buttonOnClick!.call(button, mockClickEvent);

  expect(mockClickHandler).not.toHaveBeenCalled();
  expect(mockPreventDefault).toHaveBeenCalledTimes(1);
});

it("allows a second click after a second to execute the button a second time when preventDoubleClick is true", () => {
  const mockClickHandler = jest.fn();

  const wrapper = mount(
    <Button preventDoubleClick onClick={mockClickHandler}>
      Testchildren
    </Button>
  );

  const button = wrapper.find("button");
  const buttonOnClick = button.props().onClick;

  expect(buttonOnClick).toBeDefined();

  const mockPreventDefault = jest.fn();
  const mockClickEvent = {
    ...jest.fn()(),
    target: button.getDOMNode(),
    preventDefault: mockPreventDefault,
  };

  //eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  buttonOnClick!.call(button, mockClickEvent);

  expect(mockClickHandler).toHaveBeenCalledTimes(1);
  expect(mockClickHandler).toHaveBeenCalledWith(mockClickEvent);
  expect(mockPreventDefault).not.toHaveBeenCalled();

  mockClickHandler.mockReset();

  jest.advanceTimersByTime(1000);

  //eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  buttonOnClick!.call(button, mockClickEvent);

  expect(mockClickHandler).toHaveBeenCalledTimes(1);
  expect(mockClickHandler).toHaveBeenCalledWith(mockClickEvent);
  expect(mockPreventDefault).not.toHaveBeenCalled();
});

it("prevents a click from executing the button when disabled is true", () => {
  const mockClickHandler = jest.fn();

  const wrapper = mount(
    <Button preventDoubleClick onClick={mockClickHandler} disabled>
      Testchildren
    </Button>
  );

  const button = wrapper.find("button");
  const buttonOnClick = button.props().onClick;

  const mockPreventDefault = jest.fn();
  const mockClickEvent = {
    ...jest.fn()(),
    target: button.getDOMNode(),
    preventDefault: mockPreventDefault,
  };

  //eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  buttonOnClick!.call(button, mockClickEvent);

  expect(mockClickHandler).not.toHaveBeenCalled();
  expect(mockPreventDefault).toHaveBeenCalledTimes(1);
});
