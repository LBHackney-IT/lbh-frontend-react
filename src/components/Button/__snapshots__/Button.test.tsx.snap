// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with all props 1`] = `
<button
  aria-disabled={true}
  className="govuk-button lbh-button class1 class2 govuk-button--start"
  data-prevent-double-click={true}
  disabled={true}
  id="1234"
  name="Testname"
  onClick={[Function]}
  type="button"
>
  Testchildren
  <svg
    className="govuk-button__start-icon"
    focusable="false"
    height="19"
    role="presentation"
    viewBox="0 0 33 40"
    width="17.5"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 0h13l20 20-20 20H0l20-20z"
      fill="currentColor"
    />
  </svg>
</button>
`;

exports[`renders correctly without optional props 1`] = `
<button
  className="govuk-button lbh-button"
  onClick={[Function]}
>
  Testchildren
</button>
`;
